### PR TITLE
Adds --skip flag

### PR DESCRIPTION
**Pull request by JayjeetAtGithub**
Addresses #626 , #621, #622

### CODE REVIEW

**Comment by JayjeetAtGithub** [comment]
@ivotron Please review this.

File: cli/popper/parser.py [hunk 1]
```diff
@@ -92,9 +94,14 @@ def complete_graph(self):
...
 
             for n in a_block.get('needs', []):
-                if not self.workflow['action'][n].get('next', None):
-                    self.workflow['action'][n]['next'] = set()
-                self.workflow['action'][n]['next'].add(name)
+                try:
+                    if not self.workflow['action'][n].get('next', None):
+                        self.workflow['action'][n]['next'] = set()
+                    self.workflow['action'][n]['next'].add(name)
+                except KeyError:
```
Comment (ivotron) [inline cli/popper/parser.py]: we can avoid catching the exception by doing the removal of skipped tasks to the end of the constructor (i.e. invoke `remove_skipped_actions` after `complete_graph`). When we do that, we can check whether the removal is leaving the graph in an "unrunnable" state and raise the exception notifying about it

**Comment by JayjeetAtGithub** [comment]
@ivotron  Please have a look. I have made the changes.

File: ci/test/skip [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        "curl bitbucket",
+        "curl medium",
+        "curl youtube",
+        "curl twitter",
+    ]
+}
+EOF
+
+! popper run --skip 'curl xyz'
+! popper run --skip 'curl google' --skip 'curl facebook'
```
Comment (ivotron) [inline ci/test/skip]: Sorry @jayjeetatgithub, I think I did not describe the definition of a "gap" properly. A gap is present when an action depends on a previous stage (one or more actions), but no actions for that stage are available in the workflow. So for example, for this workflow:

```hcl
workflow "wf" {
  resolves = "d"
}

action "a" {
  uses = "sh"
}

action "b" {
  needs = "a"
  uses = "sh"
}

action "c" {
  needs = "a"
  uses = "sh"
}

action "d" {
  needs = ["b", "c"]
  uses = "sh"
}
```

If we say `popper run --skip b` the workflow should run OK. In this case, we remove `b` from the workflow, and remove it as a dependency from `d` (and as a `next` item in `a`'s list of forward edges). This also should work if we do `popper run --skip c`.

If we skip `b` and `c` though, an error will be thrown because `a` cannot reach `d` (because there is a gap).

Comment (JayjeetAtGithub) [reply ci/test/skip]: Thanks a lot. @ivotron.

Comment (JayjeetAtGithub) [reply ci/test/skip]: @ivotron Can we place the `remove_skip_actions()` just before `complete_graph()` ? In this way we can leverage `complete_graph()`'s capability to find out unreachable actions. Does this sound correct ?

Comment (ivotron) [reply ci/test/skip]: the goal of the `complete_graph` function is to add forward-going edges and the root. We can add another `check_for_unreachable_nodes` function, or something alike. The goal being to have a function do a single thing to improve readability and maintainability (see [my comment](https://github.com/systemslab/popper/pull/650#discussion_r286936204) on #650).

Comment (JayjeetAtGithub) [reply ci/test/skip]: Thanks @ivotron.

Comment (JayjeetAtGithub) [reply ci/test/skip]: @ivotron  Can the order of calling the functions be like this ? 
```
self.remove_skip_actions()
self.complete_graph()
self.check_for_unreachable_nodes()`
```
In this way, we first remove the actions to be skipped, then complete the graph, and finally check for gaps? Otherwise if `remove_skip_actions()` is after `complete_graph()`, we have to again manipulate the graph.

Comment (ivotron) [reply ci/test/skip]: Sorry @JayjeetAtGithub, I should have expressed better what I have in mind.

The idea of running `complete_graph` first is to have an in-memory representation of what the `.workflow` file contains. It would simplify the code if we can have the `Workflow` object be immutable after it is constructed. Then we can apply distinct operations to the graph that return new workflow objects, for example:

 1. filter all actions but one. This is what we do when we are given an action argument.
 2. get the path from the root to a particular action (used when `--with-dependencies` is given).
 3. remove nodes (when `--skip` is given)

Each of the above can be implemented in such a way that they return a new workflow object that is the result of applying the operation to the complete workflow. This would be a functional-ish approach to manipulating the graph, as a way of making it more readable/maintainable. Then we can add other functions that do things like checking for unreachable nodes, node generators (to support distinct ways of traversing the graph), etc. 

The idea is to make the `Workflow` object a generic abstraction that can potentially be reused in multiple scenarios. For example, running the workflow multiple times (but with distinct objectives each time) in a single `popper run` invocation; e.g. run just an action first, then a particular path, then the entire workflow, etc. If we write things in such a way that the original workflow data is lost (e.g. by permanently removing nodes out of the graph), then we won't be able to implement this. Another scenario would be when popper is used as a library (e.g. in a web service), where it would be undesirable to have to read a `.workflow` file multiple times given that the file has not been modified.

So in summary, we can refactor the code so that:

  * the `Workflow`'s constructor only builds the graph, as given in the `.workflow` file.
  * the above means we need to refactor `complete_graph` so that it only completes the graph and does not do any check for unreachable nodes. 
  * add operations for the three things mentioned above
  * add a `check_for_unreachable_nodes`
  * any other thing I'm missing

what do you think?

Comment (JayjeetAtGithub) [reply ci/test/skip]: @ivotron Sorry for giving you all the trouble. I now understand what needs to be done completely.
I would merge #621, #622 in this pr. Thanks a lot !

Comment (JayjeetAtGithub) [reply ci/test/skip]: @ivotron Does removal of `a` block in the above workflow, leads to unreachable node `b`, `c`,`d` ?

Comment (JayjeetAtGithub) [reply ci/test/skip]: @ivotron Also, currently we show a warning for unreachable actions. So would that  case be changed to a fail, and occur through `check_unreachable_nodes()`?

Comment (JayjeetAtGithub) [reply ci/test/skip]: @ivotron I have done some refactoring according to your review, and solved issues #621, #626, #622 through this pr. Please review. Thanks.

File: cli/popper/parser.py [hunk 3]
```diff
@@ -198,53 +203,110 @@ def normalize(self):
...
-        for line in self.workflow_content:
+        for line in self._workflow_content:
             line = line.strip()
             if line.startswith('action '):
                 acount += 1
         if parsed_acount != acount:
             log.fail('Duplicate action identifiers found.')
+
+
+def validate_workflow(workflow):
```
Comment (ivotron) [inline cli/popper/parser.py]: this can also be a member function, or is there a particular reason this is not a member function? Also, `validate_workflow` might be a bit ambiguous. We could rename this to `find_unreachable_actions` or `check_for_unreachable_actions`

File: ci/test/skip [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+action "d" {
+  needs = ["b1", "c"]
+  uses = "sh"
+  args = "ls"
+}
+EOF
+
+! popper run
```
Comment (ivotron) [inline ci/test/skip]: we could `grep` the output to see if the command is failing due to the reasons we expect it should fail for.

File: ci/test/skip (hunk 2 above)
Comment (ivotron) [reply ci/test/skip]: awesome, thanks a lot!! I've left a review

Comment (ivotron) [reply ci/test/skip]: > @ivotron Also would the new functions like `remove_skip_actions()`, `check_unreachable_nodes()` etc that would be implemented be part of `Workflow` class, or seperate functions ? In the first case, the function call would look something like `new_wf = old_wf.remove_skip_actions()` and in the second one it will be like `new_wf = remove_skip_actions(old_wf)`.
> 
> Which approach would be better ?

@JayjeetAtGithub sorry, I missed this question. I'd vote for the latter (i.e. as static members of the class).

File: cli/popper/parser.py [hunk 5]
```diff
@@ -198,53 +203,110 @@ def normalize(self):
...
+    _traverse(workflow.root, visited, workflow)
+
+    unreachable = actions - visited - skipped
+    if unreachable:
+        log.warn('Actions {} are unreachable.'.format(', '.join(unreachable)))
+
+
+def skip_actions(workflow, skip_list):
+    """Removes the actions to be skipped from the workflow graph and
+    return a new `Workflow` object.
```
[the remark this answers is not in the document]
Comment (JayjeetAtGithub) [reply cli/popper/parser.py]: Currently it looks like the workflow was getting updated from debugging when `popper run --skip b` is ran.
```
Before:  dict_items([('a1', {'uses': 'sh', 'args': ['ls'], 'name': 'a1', 'next': {'b', 'c'}}), ('a2', {'uses': 'sh', 'args': ['ls'], 'name': 'a2', 'next': {'b', 'c'}}), ('b', {'needs': ['a1', 'a2'], 'uses': 'sh', 'args': ['ls'], 'name': 'b', 'next': {'d'}}), ('c', {'needs': ['a1', 'a2'], 'uses': 'sh', 'args': ['ls'], 'name': 'c', 'next': {'d'}}), ('d', {'needs': ['b', 'c'], 'uses': 'sh', 'args': ['ls'], 'name': 'd'})])
After :  dict_items([('a1', {'uses': 'sh', 'args': ['ls'], 'name': 'a1', 'next': {'c'}}), ('a2', {'uses': 'sh', 'args': ['ls'], 'name': 'a2', 'next': {'c'}}), ('b', {'needs': ['a1', 'a2'], 'uses': 'sh', 'args': ['ls'], 'name': 'b', 'next': {'d'}}), ('c', {'needs': ['a1', 'a2'], 'uses': 'sh', 'args': ['ls'], 'name': 'c', 'next': {'d'}}), ('d', {'needs': ['b', 'c'], 'uses': 'sh', 'args': ['ls'], 'name': 'd'})])
```
@ivotron Please verify.
But now when we move it in the `Workflow` class we need to use `deepcopy()`.

[the remark this answers is not in the document]
Comment (ivotron) [reply cli/popper/parser.py]: yes, we need to operate on a copy, otherwise the internal dict gets modified when remove skipped actions. Is that what you mean?

[the remark this answers is not in the document]
Comment (JayjeetAtGithub) [reply cli/popper/parser.py]: @ivotron  Yeah, we need to stop the original workflow from mutated.

File: cli/popper/gha.py [hunk 6]
```diff
@@ -26,11 +26,10 @@ class WorkflowRunner(object):
     """A GHA workflow runner.
     """
 
-    def __init__(self, wfile, workspace, dry_run,
+    def __init__(self, wfile, skip, workspace, dry_run,
```
[the remark this answers is not in the document]
Comment (JayjeetAtGithub) [reply cli/popper/gha.py]: @ivotron So would it be something like this ? Like in `cmd_run.py`,

```
wf = Workflow(wfile) // Initial graph generation and syntax validation
wf_runner = WorkflowRunner(wf) // All workflow manipulations happen here
wf_runner.run() // Run the workflow
```
One doubt, Where to pass all the arguments like `parallel`,  `dry_run`, `reuse` (In step 2 or step 3)?
Thanks!

[the remark this answers is not in the document]
Comment (ivotron) [reply cli/popper/gha.py]: I'd vote for passing that info to the WorkflowRunner (2)

File: ci/test/skip (hunk 4 above)
Comment (JayjeetAtGithub) [reply ci/test/skip]: @ivotron Should we warn / fail for unreachable nodes ?

Comment (ivotron) [reply ci/test/skip]: fail when `--skip` is passed (as this is likely a result of removing the skipped actions from the workflow); warn in the remaining scenarios.

**Comment by JayjeetAtGithub** [comment]
@ivotron Please have a look !

**Comment by ivotron** [comment]
thanks a lot @JayjeetAtGithub !
